graph, benchdnn: support implicit causal mask in graph API (#2330)
oneapi-src/oneDNN: doc/graph/fusion_patterns/sdpa.md

   MatMul with a scaling factor. It can be constructed by [Multiply](@ref dev_guide_op_multiply)
   or [Divide](@ref dev_guide_op_divide) operation in Graph API. The scaling
   factor is given by users as an input of SDPA. \f$\sqrt{d_k}\f$ in the formula

>    is not considered part of the SDPA pattern because it is constant.

Reviewer: ```suggestion
   is not considered as a part of the SDPA pattern because it is a constant.
```

3. The Mask node is optional and is used to apply an attention mask to the
   output of the previous Scale node. There are two types of masks that can
   be applied:

   1. Explicit user-generated mask: You can explicitly create a mask tensor
   and pass it to the library for the computation of SDPA. In this case, mask
   can be constructed by [Add](@ref dev_guide_op_add)
   or [Select](@ref dev_guide_op_select) operation in Graph API for different
   mask policies (for example, causal mask or padding mask). When the
   Add operation is used to apply the mask, the input mask is usually an upper
   triangular matrix with all the elements above the diagonal filled with
   `-inf` and zeroes elsewhere. The `-inf` entries will become zero probability
   after Softmax is applied in the next step.
   Alternately, a Select operation may be used. In this case, the

>    input is a boolean tensor (for example, with `true` on and below the

Reviewer: ```suggestion
   input is a boolean tensor (for example, with the boolean value set to  `true` below the
```

   diagonal, and `false` above the diagonal). A `false` element in the mask
   forces the corresponding element of the scaled output to `-inf`, while a
   `true` element leaves it unchanged.

   ![SDPA-mask-1](images/sdpa-mask-1.png) ![SDPA-mask-2](images/sdpa-mask-2.png)

>    2. Implicit library-generated mask: You can ask the library to generate

Reviewer: ```suggestion
   2. Implicit library-generated mask: You can use the operations in the library to generate
```

   softmax primitives. The reference implementation requires memory to store the
   intermediate results of the dot products between Query and Key which takes
   \f$O(S^2)\f$ memory. It may lead to out-of-memory error when computing long

>    sequence length input on platforms with limited memory. For implicit causal

Reviewer: ```suggestion
   sequence length input on platforms with limited memory. For an implicit causal
```
